onnxruntime/core/providers/qnn/builder/opbuilder/pad_op_builder.cc

```diff
@@ -0,0 +1,247 @@
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      case QNN_DATATYPE_SFIXED_POINT_16: {
+        auto int16_span = ReinterpretAsSpan<const int16_t>(gsl::make_span(unpacked_tensor));
+        constant_value = DequantizeValue(int16_span.data()[0],
+                                         input_info.quant_param.scaleOffsetEncoding.offset,
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      case QNN_DATATYPE_SFIXED_POINT_32: {
+        auto int32_span = ReinterpretAsSpan<const int32_t>(gsl::make_span(unpacked_tensor));
+        constant_value = DequantizeValue(int32_span.data()[0],
+                                         input_info.quant_param.scaleOffsetEncoding.offset,
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      case QNN_DATATYPE_UFIXED_POINT_8: {
+        constant_value = DequantizeValue(unpacked_tensor.data()[0],
+                                         input_info.quant_param.scaleOffsetEncoding.offset,
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      case QNN_DATATYPE_UFIXED_POINT_16: {
+        auto uint16_span = ReinterpretAsSpan<const uint16_t>(gsl::make_span(unpacked_tensor));
+        constant_value = DequantizeValue(uint16_span.data()[0],
+                                         input_info.quant_param.scaleOffsetEncoding.offset,
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      case QNN_DATATYPE_UFIXED_POINT_32: {
+        auto uint32_span = ReinterpretAsSpan<const uint32_t>(gsl::make_span(unpacked_tensor));
+        constant_value = DequantizeValue(uint32_span.data()[0],
+                                         input_info.quant_param.scaleOffsetEncoding.offset,
+                                         input_info.quant_param.scaleOffsetEncoding.scale);
+        break;
+      }
+      default:
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Type not supported for Pad constant_value.");
+    }
+    constant_value_qnn_scalar.floatValue = constant_value;
+  } else {  // constant_value is non-quantized
+    constant_value_qnn_scalar.dataType = input_info.qnn_data_type;
+    switch (input_info.qnn_data_type) {
+      case QNN_DATATYPE_UINT_8: {
+        constant_value_qnn_scalar.uint8Value = unpacked_tensor.data()[0];
+        break;
+      }
+      case QNN_DATATYPE_INT_8: {
+        auto int8_span = ReinterpretAsSpan<const int8_t>(gsl::make_span(unpacked_tensor));
+        constant_value_qnn_scalar.int8Value = int8_span.data()[0];
+        break;
+      }
+      case QNN_DATATYPE_INT_16: {
+        auto int16_span = ReinterpretAsSpan<const int16_t>(gsl::make_span(unpacked_tensor));
```

Comment:
This assumes that the initializer data is stored using the same endianess as the host machine running QNN EP. Otherwise, this reinterpret cast does not yield the correct results.  

IIRC, initializer data is always little-endian, so this would not work on big-endian machines. Is this a concern going forward?

Reply:
I don't see similar code to handle this in any EPs other than our QNN EP. I suppose it's not a problem.

---
In reply to: [1325194696](https://github.com/microsoft/onnxruntime/pull/17508#discussion_r1325194696) [](ancestors = 1325194696)

Reply:
According to ONNX proto, initializers are little-endian. Therefore, I suppose the question is this: Is QNN EP expected to run on big-endian machines? 

If the answer is yes, then I think we have two different options:
1. handle it properly here using `utils::ReadLittleEndian`.
2.  Or, just flat out have some endian check early in qnn_execution_provider that refuses to run for big-endian machines. I've seen ORT code such as:
```C++
if constexpr (endian::native != endian::little) {
  ORT_THROW("Big endian not supported");
}
```

I think just using `utils::ReadLittleEndian` is the best option, and the code would not be any more complicated:
```C++
case QNN_DATATYPE_INT_16:
  ORT_RETURN_IF_ERROR(onnxruntime::utils::ReadLittleEndian(gsl::make_span(unpacked_tensor), gsl::make_span(&contant_value_qnn_scalar.int16Value, 1)));
  break;
```

Reply:
We discussed offline and decided to handle this separately as the need arises. UnpackTensor from tensorprotoutils also handles endianness and seems like it should be used in QNN EP.

Reply:
Turns out that `UnpackInitializerData` already calls UnpackTensor, so I was wrong that we needed to handle it again here.